File: DekafMain/src/sql/TextWalker.java
package org.jetbrains.dekaf.sql;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.regex.Matcher;
import java.util.regex.Pattern;



/**
 * Text with a position pointer. Also counts line numbers.
 * @author Leonid Bushuev from JetBrains
 */
final class TextWalker implements Cloneable {

  @NotNull
  private final String myText;

  private final int myLength;

  /**
   * Offset from the beginning of the text.
   */
  private int myOffset = 0;

  /**
   * Row number, starts with 1.
   */
  private int myRow = 1;

  /**
   * Number of positions inside the row, starts with 1.
   */
  private int myPos = 1;

  /**
   * Current character.
   */
  private char myChar;



  TextWalker(@NotNull final String text) {
    myText = text;
    myLength = myText.length();
    myChar = myLength > 0 ? myText.charAt(0) : '\0';
  }


  @NotNull
  public TextPointer getPointer() {
    return new TextPointer(myOffset, myRow, myPos);
  }


  public void setPointer(@NotNull final TextPointer pointer) {
    myOffset = pointer.offset;
    myRow = pointer.row;
    myPos = pointer.pos;
    myChar = myLength > 0 ? myText.charAt(0) : '\0';
  }


  public void next() {
    if (myOffset >= myLength) throw new IllegalStateException("End of Text");

    if (myChar == '\n') {
      myRow++;
      myPos = 1;
    }
    else {
      myPos++;
    }

    myOffset++;
    myChar = myOffset < myLength ? myText.charAt(myOffset) : '\0';
  }


  public String popRow() {
    int offset1 = myOffset,
        row1 = myRow;
    while (row1 == myRow && !isEOT()) next();
    return myText.substring(offset1, myOffset);
  }


  @Nullable
  public Matcher skipToPattern(@NotNull final Pattern pattern) {
    final Matcher m = pattern.matcher(myText);
    final boolean found = m.find(myOffset);
    if (found) {
      int goal = m.start();
      while (myOffset < goal) next();
      return m;
    }
    else {
      while (myOffset < myLength) next();
      return null;
    }
  }


  public void skipToOffset(int newOffset) {
    if (newOffset < myOffset) throw new IllegalArgumentException("Attempted to skip back");
    if (newOffset > myLength) throw new IllegalArgumentException("Attempted to skip out of the end of text");

    int distance = newOffset - myOffset;
    for (int i = 1; i <= distance; i++) next();
  }


  @NotNull
  public String getText() {
    return myText;
  }


  public int getLength() {
    return myLength;
  }


  public int getOffset() {
    return myOffset;
  }


  public int getRow() {
    return myRow;
  }


  public int getPos() {
    return myPos;
  }


  public char getChar() {
    return myChar;
  }

  public char getNextChar() {
    int k = myOffset + 1;
    return k < myLength ? myText.charAt(k) : '\0';
  }


  public boolean isEOT() {
    return myOffset >= myLength;
  }


  @SuppressWarnings("CloneDoesntCallSuperClone")
  public TextWalker clone() {
    TextWalker clone = new TextWalker(myText);
    clone.myOffset = myOffset;
    clone.myRow = myRow;
    clone.myPos = myPos;
    clone.myChar = myChar;
    return clone;
  }


}

File: dev/dekaf-main/src/org/jetbrains/dekaf/text/TextWalker.kt
package org.jetbrains.dekaf.text

import java.util.regex.Matcher
import java.util.regex.Pattern


/**
 * Text with a position pointer. Also counts line numbers.
 * @author Leonid Bushuev
 */
class TextWalker(val text: String) : Cloneable {

    /**
     * The length of the text.
     */
    val length: Int = text.length

    /**
     * Current offset from the beginning of the text.
     */
    var offset = 0
        private set

    /**
     * Current row number, starts with 1.
     */
    var row = 1
        private set

    /**
     * Current number of positions inside the row, starts with 1.
     */
    var pos = 1
        private set

    /**
     * Current character.
     */
    var char: Char = if (text.isNotEmpty()) text[0] else '\u0000'
        private set

    /**
     * Current text pointer.
     */
    var pointer: TextPointer
        get() = TextPointer(offset, row, pos)
        set(pointer) {
            offset = pointer.offset
            row = pointer.row
            pos = pointer.pos
            char = if (length > 0) text[0] else '\u0000'
        }


    operator fun next() {
        if (offset >= length) throw IllegalStateException("End of Text")

        if (char == '\n') {
            row++
            pos = 1
        }
        else {
            pos++
        }

        offset++
        char = if (offset < length) text[offset] else '\u0000'
    }


    fun popRow(): String {
        val offset1 = offset
        val row1 = row
        while (row1 == row && !isEOT) next()
        return text.substring(offset1, offset)
    }


    fun skipToPattern(pattern: Pattern): Matcher? {
        val m = pattern.matcher(text)
        val found = m.find(offset)
        if (found) {
            val goal = m.start()
            while (offset < goal) next()
            return m
        }
        else {
            while (offset < length) next()
            return null
        }
    }


    fun skipToOffset(newOffset: Int) {
        if (newOffset < offset) throw IllegalArgumentException("Attempted to skip back")
        if (newOffset > length) throw IllegalArgumentException("Attempted to skip out of the end of text")

        val distance = newOffset - offset
        for (i in 1..distance) next()
    }

    val nextChar: Char
        get() {
            val k = offset + 1
            return if (k < length) text[k] else '\u0000'
        }


    val isEOT: Boolean
        get() = offset >= length


    public override fun clone(): TextWalker {
        val clone = TextWalker(text)
        clone.offset = offset
        clone.row = row
        clone.pos = pos
        clone.char = char
        return clone
    }


}
